EIPS/pairings.md

```diff
@@ -0,0 +1,141 @@
+
+## Specification
+
+Add a precompiled contracts for a bilinear function on groups on the elliptic curve "alt_bn128". We will define the precompiled contract in terms of a discrete logarithm. The discrete logarithm is of course assumed to be hard to compute, but we will give an equivalent specification that makes use of elliptic curve pairing functions which can be efficiently computed below.
+
+Address: 0x8
+
+For a cyclic group `G` (written additively) of prime order q let `log_P: G -> F_q` be the discrete logarithm on this group with respect to a generator `P`, i.e. `log_P(x)` is the integer `n` such that `n * P = x`.
+
+The precompiled contract is defined as follows, where the two groups `G_1` and `G_2` and their generators `P_1` and `P_2` are defined below (they have the same order `q`):
+
+```
+Input: (a1, b1, a2, b2, ..., ak, bk) from (G_1 x G_2)^k
+Output: If the length of the input is incorrect or any of the inputs are not elements of
+        the respective group or are not encoded correctly, the call fails.
+        Otherwise, return one if
+        log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0
+        (in F_q) and zero else.
+```
+
+Note that `k` is determined from the length of the input. `k == 0` is valid and results in returning one.
+
+In order to check that an input is an element of `G_1`, verifying the encoding of the coordinates and checking that they satisfy the curve equation (or is the encoding of infinity) is sufficient. For `G_2`, in addition to that, the order of the element has to be checked to be equal to the group order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617`.
+
+### Definition of the groups
+
+The groups `G_1` and `G_2` are cyclic groups of prime order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617` on the elliptic curve `alt_bn128` defined by the curve equation
+`Y^2 = X^3 + 3`.
+
+The group `G_1` is a cyclic group on the above curve over the field `F_p` with `p = 21888242871839275222246405745257275088696311157297823662689037894645226208583` with generator `P1 = (1, 2)`.
+
+The group `G_2` is a cyclic group on the same elliptic curve over a different field `F_p^2 = F_p[i] / (i^2 + 1)` (p is the same as above) with generator
+```
+P2 = (
+  11559732032986387107991004021392285783925812861821192530917403151452391805634 * i +
+  10857046999023057135944570762232829481370756359578518086990519993285655852781,
+  4082367875863433681332203403145435568316851327593401208105741076214120093531 * i +
+  8495653923123431417604973247489272438418190587263600148770280649306958101930
+)
+```
+
+
+### Encoding
+
+Elements of `F_p` are encoded as 32 byte big-endian numbers. An encoding value of `p` or larger is invalid.
+
+Elements `a * i + b` of `F_p^2` are encoded as two elements of `F_p`, `(a, b)`.
```

Comment:
Currently, the text is not very specific about how `k` is determined.  I think it makes sense to mention:
* pairs are encoded as concatenation of bytes, and/or
* `F_p^2` is encoded into 64 bytes
* `k` is the length of the input divided by 192
* when the input length is not divisible by 192, the input is invalid.